### DISCUSSION TITLE
Allow pixel perfect translations with animate tool

### DISCUSSION THREAD

**Discussion by Israro** [opentoonz/opentoonz#6575] [Ideas]
I have a problem with the animate tool. It happens to me with imported images and also using the raster (any bitmap image). When you move the image with the animate tool it deforms or creates new blurry pixels around it. The result is not visible in the workspace but if you look at the preview or render the work.
I understand that a bitmap image would be deformed by altering its size or perhaps by rotating it, but I find it strange that it happens only with moving it. This prevents me from making animations even a simple interpolation of moving an object from one point to another.
I do not know if this is something normal or I'm doing something wrong. I hope you can clarify my doubts.
And excuse my English, I do not speak the language well but I hope it is understood.

![01](https://user-images.githubusercontent.com/22494303/35614765-6f094352-0670-11e8-96d2-bce3075fd55e.png)
# So it looks without making any change:
![02](https://user-images.githubusercontent.com/22494303/35614766-6f2d75c4-0670-11e8-9760-52760a320474.png)
![03](https://user-images.githubusercontent.com/22494303/35614768-6f504b4e-0670-11e8-8579-485a3dfe5fe7.png)
# And this the result after moving the image with the animate tool
![04](https://user-images.githubusercontent.com/22494303/35614769-6f76ede4-0670-11e8-982f-bb39eb741c87.png)

The test was performed on two different computers and the result is the same.
The specifications are:

Laptop dell latitude E6530
Intel Core i5-3340 2.70 Ghz
4gb ram
intel HD Grafics 4000
windows 7

and a desktop HP compaq 6200 Pro
intel pentium G630 2.70 Ghz
4 Gb ram
AMD Radeon HD 6450
windows 10

**Comment by Lovewyrm** [comment]
The blur is caused by fractions in the pixel position of the raster object when animating it via keyframes that have smooth interpolation, like ease in and out, etc.

In other words, if the position of the object is not a "whole" pixel, then it will be interpolated/resampled as per the render/preview sampling settings.

One way of dealing with it is to change the "Resample Balance" in the "output settings" to nearest neighbor, aka "closest pixel".

It's not perfect but its the most direct approach. 
I think a fix would be an option that rounds pixel positions up and down.
Edit:
Rounds it only for previewing/rendering.
Aside from that, it could probably keep its floating point value in the keyframe table itself. 
Might be less code change required that way.

Maybe like "if column is not vector, interpret movement/scaling/rotation data as properly rounded integer instead of double"

**Comment by RodneyBaker** [comment]
@Lovewyrm It looks like you are on to something here.
I like the idea of keeping the floating point values for precision and rounding upon preview/render.

The fractional interpolations of pixels surely must effect other areas of Opentoonz processing as well which likely relates to issues that appear to be related to antialiasing and such.

Nice.

**Comment by ghost** [comment]
This is now a feature request.

**Comment by RodneyBaker** [comment]
There are a number of requests related to 'pixel perfect' transforms open presently.
Transferring this to Discussions to aid in identifying the best path forward.
